### PR TITLE
CMR-8649: Update api doc for the generic association

### PR DESCRIPTION
**Pull request by eudoroolivares2016**
* Adds documentation for Generic associations

### CODE REVIEW

File: search-app/docs/api.md [hunk 1]
```diff
@@ -5405,3 +5412,152 @@ Example un-healthy response body:
   }
 }
 ```
+#### <a name="generic-association"></a> Associate concepts
+
+A concept of any type can be associated to another concept of any type using the association API. Note this type of association is functionally the same as associations of tools, services, and variables to collections but, uses a different API route and is applicable to all possible concept relationships. In such fashion, any concept can be associated to any other concept, including concepts of the same type i.e. associate a variable to another variable. Associations can also include an optional data payload the purpose of which is to describe the association itself. Associations which do not initially have association details, may have details added simply by providing association details as a subsequent revision. A concept cannot be associated to itself.
```
Comment (jceaser) [inline search-app/docs/api.md:5417]: some text should be given on payload overwriting existing values on updates.

Also, I don't know about the comma after "to collections but,".

File: search-app/docs/api.md [hunk 2]
```diff
@@ -5405,3 +5412,152 @@ Example un-healthy response body:
...
+
+__Example__
+
+```
+curl -v -XPOST -H "Authorization: Bearer XXXXX" -H "Content-Type:application/json" "%CMR-ENDPOINT%/associate/GRD1200000000-PROV1" -d '[{"concept_id": "S1200000001-PROV1"}]'
```
Comment (jceaser) [inline search-app/docs/api.md:5422]: remove -v from example.

File: search-app/docs/api.md [hunk 3]
```diff
@@ -5405,3 +5412,152 @@ Example un-healthy response body:
...
+[{
+  "generic_association" : {
+    "concept_id" : "GA1200000000-CMR",
+    "revision_id" : 3
+  },
+  "associated_item" : {
+    "concept_id" : "S1200000001-PROV1"
+  }}]
+```
+
```
Comment (jceaser) [inline search-app/docs/api.md:5435]: give a "payload" example. and don't we call it just "data" in the json? Maybe we should not call it "payload" in the description as that is more of an internal title.

File: search-app/docs/api.md [hunk 4]
```diff
@@ -5405,3 +5412,152 @@ Example un-healthy response body:
...
+#### <a name="generic-multi-association"></a> Associate Multiple Concepts Simultaneously
+Multiple associaitons with or without association details can be made to a specified concept on a single request by passing multiple concept ids in a list form. The response returns the association cocncept and the concept being associated.
+__Example__
+```
+curl -v -XPOST -H "Authorization: Bearer XXXXX" -H "Content-Type:application/json" "%CMR-ENDPOINT%/associate/GRD1200000006-PROV1" -d '[{"concept_id":"S1200000008-PROV1"},{"concept_id": "C1200000007-PROV1"}]'
```
Comment (jceaser) [inline search-app/docs/api.md:5440]: drop -v

File: search-app/docs/api.md [hunk 5]
```diff
@@ -5405,3 +5412,152 @@ Example un-healthy response body:
...
+```
+__Example Response__
+```
+[{"generic_association":{"concept_id":"GA1200000013-CMR","revision_id":1},"associated_item":{"concept_id":"SO1200000012-PROV1"}}]%
+```
+
+#### <a name="generic-association-with-payloads"></a> Associate Concepts With Details
+__Example__
+```
+curl -v -XPOST -H "Authorization: Bearer XXXXX" -H "Content-Type:application/json" "%CMR-ENDPOINT%/associate/GRD1200000000-PROV1" -d '[{"concept_id": "S1200000001-PROV1", "data": {"field": "value", "another_field": {"XYZ": "ZYX"}}}]'
```
Comment (jceaser) [inline search-app/docs/api.md:5465]: -v

File: search-app/docs/api.md [hunk 6]
```diff
@@ -5405,3 +5412,152 @@ Example un-healthy response body:
...
+
+#### <a name="generic-association-search"></a> Searching For Associated Concepts
+Associations can be queried by searching for one of the assocaited concepts using either the JSON or UMM_JSON format. The "associations" field, contains a list of concepts associated to the concept that was searched for grouped by concept-type. The "association-details" field contains association details if they were included for the particular association. In the example case, a grid (GRD1200000000-PROV1) has been associated to another grid with association details(GRD1200000001-PROV1), as well as a service option (SO1200000001-PROV1) without any association details.
+
+__Example__
+
+```
+curl -g -H "Authorization Bearer: XXXXX" "%CMR-ENDPOINT%/grids?concept_id=GRD1200000000-PROV1"
```
Comment (jceaser) [inline search-app/docs/api.md:5487]: -g is assumed right?

**Review by jceaser — CHANGES_REQUESTED** [review]
left comments in the files

File: search-app/docs/api.md (hunk 5 above)
Comment (eudoroolivares2016) [reply search-app/docs/api.md:5465]: I had also interpreted "payload" as an internal working title, I kept it part of the markdown thinking it was useful but, it isn't used in the actual readable documentation. I can change it to data though to make everything a bit more consistent

File: search-app/docs/api.md (hunk 1 above)
Comment (eudoroolivares2016) [reply search-app/docs/api.md:5417]: I do expand on that topic in the: "Update Association details" section, perhaps it should be in the main section or deserves it's own TOC entry?

File: search-app/docs/api.md [hunk 7]
```diff
@@ -210,6 +210,13 @@ See the [CMR Client Partner User Guide](https://wiki.earthdata.nasa.gov/display/
     * [Reset the application to the initial state](#reset-the-application-to-the-initial-state)
     * [Querying caches](#querying-caches)
     * [Check application health](#check-application-health)
+  * [Associate Concepts](#generic-association)
+    * [Associate Multiple Concepts](#generic-multi-association)
+    * [Associate Concepts By Revision](#generic-association-with-revisions)
+    * [Associate Concepts With Details](#generic-association-with-details)
+    * [Search For Associations](#generic-association-search)
+    * [Disassociate Concepts](#generic-association-delete)
+    * [Disassociate Multiple Concepts](#generic-association-multi-delete)
```
Comment (sxu123) [inline search-app/docs/api.md:219]: I don't think it's necessary to have a section for multiple association and dissociations.  People are already familiar with a concept being associated/dissociated with multiple concepts. If you want to associate with only one concept, that's just special case for the multiple concepts. So, you just need to cover the multiple case, without calling it multiple. If the user wants to only associate one, it's their choice.

**Comment by eudoroolivares2016** [comment]
Closing, because after some revision we are now adding this documentation to the CMR-8673 ticket. https://github.com/nasa/Common-Metadata-Repository/pull/1756

**Pull request closed without merging**